easyrsa3/easyrsa

```diff
@@ -1749,8 +1734,19 @@ Please update openssl-easyrsa.cnf to the latest official release."
 			die "Failed to read X509-type $crt_type"
 
 		# Support a dynamic CA path length when present:
-		[ "$crt_type" = "ca" ] && [ "$EASYRSA_SUBCA_LEN" ] && \
-			print "basicConstraints = CA:TRUE, pathlen:$EASYRSA_SUBCA_LEN"
+		if [ "$crt_type" = "ca" ] && [ "$EASYRSA_SUBCA_LEN" ]; then
+			# Print the last occurence of basicContraints in x509-types/ca
+			# If basicContraints not defined then bail
+			# shellcheck disable=SC2016 # vars don't expand in ''
+			awkscript='/^[[:blank:]]*basicConstraints[[:blank:]]*=/ { bC=$0 }
+				END { if (length(bC) == 0 ) exit 1; print bC }'
+			basicConstraints="$(
+				awk "$awkscript" "$EASYRSA_EXT_DIR/$crt_type"
+				)" || die "\
+basicConstraints is not defined, cannot use 'pathlen'"
```

Comment:
I'm still seeing show_host() and cleanup() information in $ext_tmp :( 
`easyrsa --keep-tmp=foo1 --subca-len=99 sign-req ca ssubtCA`
```
# === Trimmed ===
#basicConstraints = CA:TRUE                                                                                             
#basicConstraints = CA:TRUE, pathlen:9                                                                                  
subjectKeyIdentifier = hash                                                                                             
authorityKeyIdentifier = keyid:always,issuer:always                                                                     
keyUsage = cRLSign, keyCertSign                                                                                         
* Temp session preserved: /home/dekeonus/devel/easyrsa/pki_root/subtCA/tmp/foo1                                         
                                                                                                                        
                                                                                                                        
EasyRSA Version Information                                                                                             
Version:     git-706r2                                                                                                  
Generated:   2022-09-27T20:22:05 AEST                                                                                   
SSL Lib:     OpenSSL 1.1.1q  5 Jul 2022                                                                                 
Git Commit:  43e5cb9af1decd6a09734b3955fea37b4be7d999                                                                   
Source Repo: https://github.com/OpenVPN/easy-rsa                                                                        
Host: git-706r2 | nix | Linux | /bin/bash | OpenSSL 1.1.1q  5 Jul 2022 
```

Reply:
This is fixed by https://github.com/OpenVPN/easy-rsa/pull/706/commits/96b3d3884722994766fa9f557886be84f265f431

Reply:
I'm still seeing the issue:
```
# === trimmed ===
#basicConstraints = CA:TRUE, pathlen:9
subjectKeyIdentifier = hash
authorityKeyIdentifier = keyid:always,issuer:always
keyUsage = cRLSign, keyCertSign
* Temp session preserved: /home/dekeonus/devel/easyrsa/pki_root/subtCA/tmp/foo3


EasyRSA Version Information
Version:     git-706-r3
Generated:   2022-09-27T21:02:38 AEST
SSL Lib:     OpenSSL 1.1.1q  5 Jul 2022
Git Commit:  96b3d3884722994766fa9f557886be84f265f431
Source Repo: https://github.com/OpenVPN/easy-rsa
Host: git-706-r3 | nix | Linux | /bin/bash | OpenSSL 1.1.1q  5 Jul 2022
```

Reply:
the `shasum easyrsa` for the unversioned easyrsa script I have here (i.e. ~VER~  ~DATE~  et.al.)
421c9ceb7d8a68f023914db35f3a423a3f713869  easyrsa

Reply:
There is a minor merge conflict but after successful merge you should have this:
```
die() {
	print "
Easy-RSA error:

$1
" 1>&2
	die_error_exit=1
	exit "${2:-1}"
} # => die()
```
`die()` does not call `show_host()`.

Then
```
cleanup() {
	if [ "${EASYRSA_TEMP_DIR_session%/*}" ] && \
		[ -d "$EASYRSA_TEMP_DIR_session" ]
	then
		# Remove temp-session or create temp-snapshot
		if [ "$EASYRSA_KEEP_TEMP" ]
		then
			# skip on black-listed directory names, with a warning
			if [ -d "$EASYRSA_TEMP_DIR/$EASYRSA_KEEP_TEMP" ]
			then
				warn "\
Prohibited value for --keep-tmp: '$EASYRSA_KEEP_TEMP'
Temporary session not preserved."
			else
				# create temp-snapshot
				keep_tmp="$EASYRSA_TEMP_DIR/tmp/$EASYRSA_KEEP_TEMP"
				mkdir -p "$keep_tmp"
				rm -rf "$keep_tmp"
				mv -f "$EASYRSA_TEMP_DIR_session" "$keep_tmp"
				print "Temp session preserved: $keep_tmp"
			fi
		fi

		# Always remove temp-session
		rm -rf "$EASYRSA_TEMP_DIR_session"
	fi

	if [ "${EASYRSA_EC_DIR%/*}" ] && [ -d "$EASYRSA_EC_DIR" ]
	then
		rm -rf "$EASYRSA_EC_DIR"
	fi

	# Remove files when build_full()->sign_req() is interrupted
	[ "$on_error_build_full_cleanup" ] && \
		rm -f "$crt_out" "$req_out" "$key_out"

	# Restore files when renew is interrupted
	[ "$on_error_undo_renew_move" ] && renew_restore_move; :
	# Restore files when rebuild is interrupted
	[ "$on_error_undo_rebuild_move" ] && rebuild_restore_move; :

	# shellcheck disable=SC3040 # In POSIX sh, set option [name] is undefined
	case "$easyrsa_host_os" in
		nix) [ -t 1 ] && stty echo ;;
		win)
			if [ "$easyrsa_win_git_bash" ]; then
				[ -t 1 ] && stty echo
			else
				set -o echo
			fi
		;;
		*) warn "Host OS undefined."
	esac

	if [ "$1" = ok ] || [ "$EASYRSA_BATCH" ] || \
		[ "$EASYRSA_SILENT" ] || [ "$EASYRSA_QUIET" ]
	then
		: # ok
	else
		print # just to get a clean line
	fi

	# Exit with error 1, if an error ocured...
	if [ "$easyrsa_error_exit" ]; then
		# Set by verify_cert() for full error-out
		exit 1
	elif [ "$1" = ok ]; then
		# if there is no error then 'cleanup ok' is called
		exit 0
	else
		# if 'cleanup' is called without 'ok' then an error occurred
		# Do not show_host() for confirm() aborted exit
		[ "$die_error_exit" ] && show_host
		exit 1
	fi
} # => cleanup()
```
`cleanup()` has the new safe version of `--keep-tmp` and only calls `show_host()` on an error sent to `die()`.  This allows for aborted `confirm()` to exit without extended error data.

Reply:
That is the code I've got here, and what I was running.

I opened the commit link https://github.com/OpenVPN/easy-rsa/commit/96b3d3884722994766fa9f557886be84f265f431 , selected browse files, used the download code link and copied the easyrsa script to my test location. and then modified the version strings by hand¹ (so that I have a reference for what code I'm running in the test terminal).

I've got modifications in my working dir, and I didn't want to git stash them.

¹ in my normal workflow I use a script to call the build-dist.sh with the local branch name as version for testing.

Reply:
Sorry that cleanup() function you listed does not match https://github.com/OpenVPN/easy-rsa/commit/96b3d3884722994766fa9f557886be84f265f431

Reply:
You have applied patches out of order.

Reply:
I was downloading the file (entire not patches) from github.
I've now, in a fresh never before used directory, performed:
`git clone https://github.com/TinCanTech/easy-rsa.git`
`git fetch origin fix-subca-len`
`git switch fix-subca-len`
`git log -1`
```
commit 43e5cb9af1decd6a09734b3955fea37b4be7d999 (HEAD -> fix-subca-len, origin/fix-subca-len)
```

deploying `easyrsa` into my test environment with that TinCanTech/fix-subca-len branch still results in show_host() and information() ending in the temp extension file.

Reply:
You need to apply this to 318e57bbbc46ce6c15acbe57696304cb3a24d73b